src/Perlang.Interpreter/Compiler/PerlangCompiler.cs

```diff
@@ -897,12 +897,22 @@ public object VisitAssignExpr(Expr.Assign expr)
                 // to ensure we get the expected operator on the C++ side (example: `int + uint` should produce a
                 // `long`, and at least one of the operands need to be widened to ensure expected semantics for an
                 // expression like `2 + 4294967295`)
-                if (expr.TypeReference.ClrType != expr.Left.TypeReference.ClrType)
+                //
+                // String types are special-cased here, since they are wrapped in std::shared_ptr and cannot be cast to
+                // e.g perlang::String directly (because this will make the C++ compiler attempt to perform a copy to an
+                // abstract, non-instantiable type).
+                if (expr.TypeReference.ClrType != expr.Left.TypeReference.ClrType && !(
+                        expr.TypeReference.ClrType == typeof(Lang.String) ||
+                        expr.TypeReference.ClrType == typeof(Lang.AsciiString) ||
+                        expr.TypeReference.ClrType == typeof(Lang.Utf8String)))
                 {
                     leftCast = expr.TypeReference.CppTypeCast;
                 }
 
-                if (expr.TypeReference.ClrType != expr.Right.TypeReference.ClrType)
+                if (expr.TypeReference.ClrType != expr.Right.TypeReference.ClrType && !(
+                        expr.TypeReference.ClrType == typeof(Lang.String) ||
+                        expr.TypeReference.ClrType == typeof(Lang.AsciiString) ||
+                        expr.TypeReference.ClrType == typeof(Lang.Utf8String)))
```

Comment:
These were added here to avoid peculiar constructs like `*s1 + *(perlang::String)perlang::ASCIIString::from_static_string(" ")) + *s2` in the generated code. The problem is that while this would be kind-of-valid in C# and Java, we can't really cast to a `perlang::String` like this because this triggers an instantiation => `perlang::String` is abstract and cannot (and should not!) ever be instantiated. Ignoring the cast for now is the simplest approach here; let's see if all the tests are still passing or if this breaks something else now...